csrc/device_lower/validation.cpp

```diff
@@ -261,6 +261,251 @@ bool isInnermost(IterDomain* base_id, IterDomain* maybe_innermost_id) {
   return !frontier.empty() && frontier.back() == maybe_innermost_id;
 }
 
+// Helper class for BlockQuantization validation
+// This class validates the following scheduling requirements for
+// BlockQuantizationOp:
+// 1. The Group ID must be derived from the innermost logical IDs
+// 2. Merge operations must combine contiguous logical IDs
+// 3. TIDx must follow the Group ID in the schedule -- that is when derived from
+// the logical domain, group ID must be inner-most, the next "inner-most" should
+// be TIDx (unless there is an ID with a unit trip count)
+class BlockQuantizationValidationHelper {
+ public:
+  BlockQuantizationValidationHelper(const TensorView* tv, IterDomain* group_id)
+      : tv_(tv), group_id_(group_id) {}
+
+  void run() {
+    // Trace back from Group ID to logical domain to see group is "inner-most"
+    traceGroupIdToLogicalDomain();
+    // Find where to start the search for TIDx
+    IterDomain* restart_traversal_from = determineThreadXTraversalStart();
+    // We start a DFS for TIDx taking the inner outputs of Split nodes first.
+    // We should not find a terminating ID that is not TIDx and has extent > 1.
+    validateThreadXFollowsGroupId(restart_traversal_from);
+  }
+
+ private:
+  // Checks if a merge operation combines contiguous IDs from the logical
+  // domain. A merge is acceptable if its inputs can be traced back to IDs in
+  // the logical domain and those IDs are contiguous.
+  bool isAcceptableMerge(Merge* merge) const {
+    const auto& logical_domain = tv_->getLogicalDomain();
+    auto ids_in_logical_domain = IterVisitor::getInputsTo({merge->out()});
```

Comment:
This isn't guaranteed to return logical IDs. It just returns IDs with no further definition.